torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -469,7 +473,22 @@ void RequestCallbackImpl::processRpc(
           [responseFuture,
            messageId,
            fromWorkerId,
-           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture)]() {
+           weak = std::weak_ptr<FutureMessage>(wrappedRpcResponseFuture),
+           threadLocalState = ThreadLocalState(),
```

Comment:
Do we need TLS state or just the dist autograd context id for now? (I am planning to eventually use TLS state for distributed profiler work, but curious if this is already needed now)

Reply:
Also, could we potentially reuse the approach taken in `record_function_ops.cpp`?
It basically does:
```
// Save and pass thread local state into the callback
  at::ThreadLocalState tls_state;
fut->addCallback([tls_state = std::move(tls_state)]() {
// restore tls_state in lambda
at::ThreadLocalStateGuard g(tls_state);
});
```

The reasoning for this is that using `ThreadLocalStateGuard` is preferred to directly accessing `ThreadLocalState` as I understand it, after some discussion with @ilia-cher

Reply:
> Do we need TLS state or just the dist autograd context id for now? (I am planning to eventually use TLS state for distributed profiler work, but curious if this is already needed now)

For now we only need the context id I think. @xush6528 also pointed out that we will need this for profiler later, so added `ThreadLocalState` as well. I think we can also remove it in this PR and leave to the profiler-related PR?

> This is basically the same here, but there we declare the tls_state outside of the lambda capture and std::move it into the capture. Not sure if there's a difference perf wise.

Should be the same I guess, as both are rvalue references?

Reply:
You can remove. I think profiler state is not needed because no more tensor operations happens in this callback.
Autograd context id is needed because getMessageWithAutograd(..) needs it.

We will need to restore both profiler state and autograd context id for python call user async function continuation.